osdconfig/api.go

```diff
@@ -0,0 +1,275 @@
+	defer manager.watchLock.Unlock()
+
+	// loop forever until watch context is alive
+	for {
+		select {
+		case <-manager.ctx.Done():
+			manager.info("context cancellation received in: kvdb watch")
+			manager.info("no longer watching for kvdb updates")
+			return
+		case wd := <-c:
+			if wd.Err != nil {
+				manager.error("error during callback execution by kvdb")
+			} else {
+				select {
+				case err := <-manager.Run(wd):
+					if err != nil {
+						manager.error("error during registered callback execution. " +
+							"Not all callbacks executed properly")
+					}
+				}
+			}
+		}
+	}
+}
+
+// close and cleanup if any
+func (manager *configManager) Close() {
+	// cancel contexts and wait for any pending tasks to complete
+	manager.localCancel()
+	manager.cancel()
+	manager.Wait()
+
+	// clean up resources
+	manager.cc = nil
+	manager.cb = nil
+	manager.opt = nil
+}
+
+// send context cancellation to any callbacks running
+func (manager *configManager) Abort() {
+	manager.info("aborting via context cancellation")
+	manager.localCancel()
+	manager.Lock()
+	manager.info("abort successful")
+	manager.Unlock()
+}
+
+// wait for manager to release lock
+func (manager *configManager) Wait() {
+	manager.Lock()
+	manager.Unlock()
+}
+
+// wrap logrus error
+func (manager *configManager) error(args ...interface{}) {
+	if manager.logLevel > LOG_NONE {
+		logrus.Error(args...)
+	}
+}
+
+// wrap logrus info
+func (manager *configManager) info(args ...interface{}) {
+	if manager.logLevel == LOG_INFO_AND_ERRORS {
+		logrus.Info(args...)
+	}
+}
+
+// GetClusterConf retrieves cluster level data from kvdb
+func (manager *configManager) GetClusterConf() (*ClusterConfig, error) {
+	// get json from kvdb and unmarshal into config
+	kvPair, err := manager.cc.Get(filepath.Join(base_key, cluster_key))
+	if err != nil {
+		return nil, err
+	}
+
+	if kvPair == nil {
+		return nil, errors.New("no data fetched from kvdb store")
+	}
+
+	config := new(ClusterConfig)
+	if err := json.Unmarshal(kvPair.Value, config); err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// SetClusterConf sets cluster config in kvdb
+func (manager *configManager) SetClusterConf(config *ClusterConfig) error {
+	if config == nil {
+		return errors.New("input is nil")
+	}
+
+	// push into kvdb
+	_, err := manager.cc.Put(filepath.Join(base_key, cluster_key), config, 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetNodeConf retrieves node config data using nodeID
+func (manager *configManager) GetNodeConf(nodeID string) (*NodeConfig, error) {
+	if len(nodeID) == 0 {
+		return nil, errors.New("node id input is empty")
+	}
+
+	// get json from kvdb and unmarshal into config
+	kvPair, err := manager.cc.Get(getNodeKeyFromNodeID(nodeID))
+	if err != nil {
+		return nil, err
+	}
+
+	if kvPair == nil {
+		return nil, errors.New("no data fetched from kvdb store")
+	}
+
+	config := new(NodeConfig)
+	if err := json.Unmarshal(kvPair.Value, config); err != nil {
```

Comment:
don't need `err :=` since err is already initialized above

Reply:
nit? :) yes, we can use previously declared err... i was just in a habit to write code this way since it keeps scope local to that `if` statement and is more idiomatic.

Reply:
done